plutus-tx/src/PlutusTx/DataMap.hs

```diff
@@ -0,0 +1,355 @@
+  Just a  -> Just (P.unsafeFromBuiltinData a)
+  Nothing -> Nothing
+
+{-# INLINEABLE lookup' #-}
+lookup' ::
+  BuiltinData ->
+  BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) ->
+  Maybe BuiltinData
+lookup' k = go
+  where
+    go xs =
+      P.matchList
+        xs
+        Nothing
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then Just (BI.snd hd)
+                  else go tl
+        )
+
+{-# INLINEABLE member #-}
+member :: forall k a. (P.ToData k) => k -> Map k a -> Bool
+member (P.toBuiltinData -> k) m = member' k (toBuiltinList m)
+
+{-# INLINEABLE member' #-}
+member' :: BuiltinData -> BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+member' k = go
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        False
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then True
+                  else go tl
+        )
+
+{-# INLINEABLE insert #-}
+insert :: forall k a. (P.ToData k, P.ToData a) => k -> a -> Map k a -> Map k a
+insert (P.toBuiltinData -> k) (P.toBuiltinData -> a) m =
+  unsafeFromBuiltinList (go (toBuiltinList m))
+  where
+    go ::
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) ->
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    go xs =
+      P.matchList
+        xs
+        (BI.mkCons (BI.mkPairData k a) nil)
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then BI.mkCons (BI.mkPairData k a) tl
+                  else BI.mkCons hd (go tl)
+        )
+
+{-# INLINEABLE delete #-}
+delete :: forall k a. (P.ToData k) => k -> Map k a -> Map k a
+delete (P.toBuiltinData -> k) m = unsafeFromBuiltinList (go (toBuiltinList m))
+  where
+    go ::
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) ->
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    go xs =
+      P.matchList
+        xs
+        nil
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then tl
+                  else BI.mkCons hd (go tl)
+        )
+
+{-# INLINEABLE singleton #-}
+singleton :: forall k a. (P.ToData k, P.ToData a) => k -> a -> Map k a
+singleton (P.toBuiltinData -> k) (P.toBuiltinData -> a) = unsafeFromBuiltinList xs
+  where
+    xs = BI.mkCons (BI.mkPairData k a) nil
+
+{-# INLINEABLE empty #-}
+empty :: forall k a. Map k a
+empty = unsafeFromBuiltinList nil
+
+{-# INLINEABLE null #-}
+null :: forall k a. Map k a -> Bool
+null = P.null . toBuiltinList
+
+{-# INLINEABLE unsafeFromList #-}
+unsafeFromList :: (P.ToData k, P.ToData a) => [(k, a)] -> Map k a
+unsafeFromList =
+  unsafeFromBuiltinList
+    . toBuiltin
+    . PlutusTx.Prelude.map (\(k, a) -> (P.toBuiltinData k, P.toBuiltinData a))
+
+{-# INLINEABLE uncons #-}
```

Comment:
Would you ever want to do this?

Reply:
Yeah, this is a bit weird, since it depends on the ordering

Reply:
The Marlowe validator needs `unsafeUncons`.

This data structure is really an association list rather than a map, and should probably be named so. For an association list, it makes a lot more sense to have an `uncons` operation.